changefeedccl: add pts management performance benchmark

aerfrei opened the pull request:
This change adds benchmarking for the behavior of the protected timestamp system. The benchmark is a roachtest that creates 200 instances of the tpcc workload and starts a changefeed watching their 200 order tables for 45 minutes.

We will use this same test to measure the performance of the pts system for multi-table feeds once they create a distinct pts record for each watched table.

Epic: CRDB-1421
Fixes: #147781
Release note: None

asg0451 commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -2657,6 +2802,23 @@ func registerCDC(r registry.Registry) {
+		CompatibleClouds: registry.AllClouds,
+		Suites:           registry.Suites(registry.Nightly),
+		Timeout:          1 * time.Hour,
+		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
```
what's the output of this test? should we be measuring something like max pts age? or maybe inspecting system.span_configurations?

KeithCh commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -2657,6 +2802,23 @@ func registerCDC(r registry.Registry) {
+		Name:             "cdc/multi-table-pts-benchmark",
+		Owner:            registry.OwnerCDC,
+		Benchmark:        false,
+		Cluster:          r.MakeClusterSpec(3, spec.CPU(16), spec.WorkloadNode()),
```
Curious how did you settle on this cluster spec? 

My guess is that it doesn't really mater as long as cluster size >= 3.

asg0451 replied:
you should use 4 nodes instead of 3 here, so there's a dedicated workload node.

andyyang890 commented on pkg/workload/bank/bank.go:
```diff
@@ -186,13 +209,17 @@ func (b *bank) Ops(
-	if err != nil {
-		return workload.QueryLoad{}, errors.CombineErrors(err, db.Close())
+	updateStmts := make([]*gosql.Stmt, b.numTables)
+	for tableIdx := 0; tableIdx < b.numTables; tableIdx++ {
```
nit: replace with for-range loop

```suggestion
	for tableIdx := range b.numTables {
```

andyyang890 commented on pkg/workload/bank/bank.go:
```diff
@@ -34,6 +34,7 @@ const (
 	defaultBatchSize    = 1000
 	defaultPayloadBytes = 100
 	defaultRanges       = 10
+	defaultNumTables    = 1
```
Can you pull the bank workload changes into its own commit with a message summarizing the changes?

andyyang890 commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -1731,6 +1731,122 @@ func runMessageTooLarge(ctx context.Context, t test.Test, c cluster.Cluster) {
 
+type multiTablePTSBenchmarkParams struct {
+	numRows   int
+	numTables int
```
super nit: put `numTables` first so it's in descending order of size

aerfrei replied:
Like size in terms of length of the string `numTables`? Should duration be between them?

andyyang890 replied:
As in granularity: table > row

andyyang890 commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -1731,6 +1731,122 @@ func runMessageTooLarge(ctx context.Context, t test.Test, c cluster.Cluster) {
+
+	startOpts := option.DefaultStartOpts()
+	startOpts.RoachprodOpts.ExtraArgs = append(startOpts.RoachprodOpts.ExtraArgs,
+		"--vmodule=changefeed=2",
+		"--vmodule=changefeed_processors=2",
+		"--vmodule=protected_timestamps=2",
```
I suspect this won't do what you want it to do, i.e. the last value of the flag will be used. You can instead pass all of them as a comma-separated list with a single flag.

andyyang890 commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -1731,6 +1731,122 @@ func runMessageTooLarge(ctx context.Context, t test.Test, c cluster.Cluster) {
+func setClusterSettingForMultiTablePTSBenchmark(db *gosql.DB) error {
+	// This is used to trigger frequent garbage collection and
+	// protected timestamp updates.
+	if _, err := db.Exec("ALTER DATABASE defaultdb CONFIGURE ZONE USING gc.ttlseconds = 1"); err != nil {
```
nit: technically not a cluster setting, maybe you could name the function `configureDBFor...` instead?

andyyang890 commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -1731,6 +1731,122 @@ func runMessageTooLarge(ctx context.Context, t test.Test, c cluster.Cluster) {
+		},
+	})
+
+	t.Status("Multi-table PTS benchmark running with jobId ", feed.jobID)
```
nit: don't capitalize status message (ditto for the other ones)

andyyang890 commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -1731,6 +1731,122 @@ func runMessageTooLarge(ctx context.Context, t test.Test, c cluster.Cluster) {
+	}
+
+	targetNames := make([]string, 0, params.numTables)
+	for i := 0; i < params.numTables; i++ {
```
nit: replace with for-range loop
```suggestion
	for i := range params.numTables {
```

andyyang890 commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -1731,6 +1731,122 @@ func runMessageTooLarge(ctx context.Context, t test.Test, c cluster.Cluster) {
+
+	// These metrics are in nanoseconds, so we are asserting that both
+	// of these latency metrics are less than 10 milliseconds.
+	ct.verifyMetrics(ctx, verifyMetricsUnderThreshold([]string{
```
Could we assert that this function returns true?

andyyang890 replied:
Ah you're right, I misread it

andyyang890 commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -1731,6 +1731,122 @@ func runMessageTooLarge(ctx context.Context, t test.Test, c cluster.Cluster) {
+	ct.verifyMetrics(ctx, verifyMetricsUnderThreshold([]string{
+		"changefeed_stage_pts_manage_latency",
+		"changefeed_stage_pts_create_latency",
+	}, 10_000_000))
```
nit: best to avoid raw numbers when possible
```suggestion
	}, float64(10*time.Millisecond)))
```

aerfrei committed:
workload: allow bank workload to run for multiple tables

Previously running the bank workload would spin up a single
bank table and make changes to it with transfers between
its rows/accounts. Now, with the num-tables argument specified,
this workload will create that many tables (e.g. bank_1 to
bank_n) and each transfer will choose a different table on
which to make a transfer between two random accounts.

This is a useful workload to be able to test multi-table
and db-level changefeeds.

Release note: None

aerfrei committed:
changefeedccl: add pts management performance benchmark

This change adds benchmarking for the behavior of the
protected timestamp system. The benchmark is a roachtest
that creates 500 instances of the bank workload and starts
a changefeed watching their 500 tables for 20 minutes.

We will use this same test to measure the performance of the
pts system for multi-table feeds once they create a distinct
pts record for each watched table.

Epic: CRDB-1421
Fixes: #147781
Release note: None

aerfrei commented:
bors cancel

aerfrei commented:
bors r=andyyang890
